core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -371,10 +363,11 @@ private RebalanceDecision decideRebalance(final ShardRouting shard) {
             final String idxName = shard.getIndexName();
             Map<String, NodeRebalanceResult> nodeDecisions = new HashMap<>(modelNodes.length - 1);
             Type rebalanceDecisionType = Type.NO;
-            String assignedNodeId = null;
+            DiscoveryNode assignedNode = null;
```

Comment:
maybe just store ModelNode here and later resolve when passing it to RebalanceDecision.